File: app/src/main/java/com/example/android/sunshine/app/index/IndexPresenter.java
package com.example.android.sunshine.app.index;

import android.util.Log;

import com.example.android.sunshine.app.shell.ShellContract;

import java.util.Date;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

/**
 * This file is a part of the Sunshine-Version-2 project.
 */
public class IndexPresenter
        implements IndexContract.Interaction, IndexContract.Synchronization {
    public static final String TAG = "mzIndexPresenter";
    private IndexContract.Display view;

    @Inject @Named("forecasts")
    List<IndexContract.Forecast> forecasts;

    @Inject
    ShellContract.Navigation go;

    @Override
    public void didPressRefresh() {
        getForecasts();
    }

    @Override
    public void didChooseForecast(int index) {
        Log.d(TAG, String.format("showing forecast #%d", index));
        // go.launch(DETAILS, Bundle.withInt("id", forecasts.get(index).id)))
    }

    @Override
    public void bind(IndexContract.Display view) {
        this.view = view;
    }

    @Override
    public void getForecasts() {
        Log.d(TAG, "fetching");
        // service.fetch(5, this::gotForecasts)
    }

    @Override
    public void gotForecasts(List<IndexContract.Forecast> xs) {
        // forecasts.clear();
        // forecasts.addAll(xs);
        Log.d(TAG, String.format("got %d items", xs.size()));
        forecasts.add(IndexWiring.Weather.create(new Date(), "this is new!", 100, 10));
        if (view != null) {
            view.refresh();
        }
    }
}

File: solshine/src/main/java/ph/codeia/solshine/index/IndexPresenter.kt
package ph.codeia.solshine.index

import android.content.SharedPreferences
import android.os.Bundle
import ph.codeia.solshine.BuildConfig
import ph.codeia.solshine.IndexState
import ph.codeia.solshine.Result
import ph.codeia.solshine.openweathermap.OwmService
import ph.codeia.solshine.shell.ShellContract
import ph.codeia.solshine.shell.ShellContract.Duration
import java.text.DateFormat
import javax.inject.Inject

/**
 * This file is a part of the Sunshine-Version-2 project.
 */
class IndexPresenter @Inject constructor(
        private val state: IndexState,
        private val go: ShellContract.Navigation,
        private val msg: ShellContract.Feedback,
        private val repository: OwmService,
        private val prefs: SharedPreferences,
        private val dateFormat: DateFormat
) : IndexContract.Interaction, IndexContract.Synchronization {

    private var view: IndexContract.Display? = null

    override fun didPressRefresh() {
        if (!state.isPending) {
            state.isStale = true
            fetchForecasts()
        }
    }

    override fun didChooseItem(index: Int) {
        msg.toast("you clicked on $index", Duration.SHORT)
        go.launch(ShellContract.DETAIL, Bundle().apply {
            @Suppress("NEW_API")  // surely a bug, these are available since API 1.
            state.items[index].let {
                putString("date", dateFormat.format(it.date))
                putString("status", it.status)
                putDouble("min", it.minTemp)
                putDouble("max", it.maxTemp)
                state.location?.let { loc ->
                    putString("location", "${loc.city}, ${loc.country}")
                }
            }
        })
    }

    override fun didPressViewMap() {
        state.location?.let {
            val (lat, long) = it.coords
            @Suppress("fix this bundle api level bug pls.")
            go.launch(ShellContract.MAP, Bundle().apply {
                putDouble("lat", lat)
                putDouble("long", long)
            })
        }
    }

    override fun bind(view: IndexContract.Display?) {
        this.view = view
    }

    override fun fetchForecasts() {
        if (state.isStale && !state.isPending) {
            state.isPending = true
            repository.fetchForecasts(
                    BuildConfig.OWM_API_KEY,
                    prefs.getString("location", "Manila"),
                    units = prefs.getString("units", "metric")
            ) {
                state.isPending = false
                when (this) {
                    is Result.Ok -> value.apply {
                        forecastsFetched(forecasts.map {
                            Weather(it.weather.category, it.date,
                                    it.temperature.min, it.temperature.max)
                        }, Location(city, country, coords))
                    }
                    is Result.Err -> msg.tell(error.toString())
                }
            }
        }
    }

    override fun forecastsFetched(
            newItems: List<IndexContract.WeatherData>,
            location: IndexContract.Location
    ) {
        msg.log("got stuff: $newItems")
        state.items.clear()
        state.items.addAll(newItems)
        state.location = location
        state.isStale = false
        go.title = location.city
        view?.refresh()
    }
}